src/crimson/osd/ec_backend.cc

```diff
@@ -19,20 +58,715 @@ ECBackend::_read(const hobject_t& hoid,
                  const uint64_t len,
                  const uint32_t flags)
 {
-  // todo
-  return seastar::make_ready_future<bufferlist>();
+  LOG_PREFIX(ECBackend::_read);
+  const auto [aligned_off, aligned_len] =
+    sinfo.offset_len_to_stripe_bounds(std::make_pair(off, len));
+  std::map<hobject_t, std::list<ec_align_t>> reads;
+  reads[hoid].emplace_back(
+    ec_align_t{aligned_off, aligned_len, flags});
+  seastar::promise<ceph::bufferlist> promise;
+  auto ret = promise.get_future();
+  objects_read_and_reconstruct(
+    reads,
+    fast_read,
+    make_gen_lambda_context<ec_extents_t &&>(
+      [hoid, off, len, promise=std::move(promise), FNAME](auto&& results) mutable {
+        ceph_assert(results.size() == 1);
+        ceph_assert(results.count(hoid) == 1);
+        auto& [err, extents] = results[hoid];
+        if (err < 0) {
+          ceph_abort_msg("implement error handling");
+          return;
+        }
+	auto range = extents.get_containing_range(off, len);
+	ceph_assert(range.first != range.second);
+	ceph_assert(range.first.get_off() <= off);
+        DEBUG("offset: {}", off);
+        DEBUG("range offset: {}", range.first.get_off());
+        DEBUG("length: {}", len);
+        DEBUG("range length: {}", range.first.get_len());
```

Comment:
Every debug line should use DEBUGDPP, we really do not want *any* log lines without the pg prefix unless absolutely necessary.